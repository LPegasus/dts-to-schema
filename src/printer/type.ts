import type * as ts from "typescript";

export interface PrintAdaptor {
  emitSchema(printSourceData: PrintSourceData): Promise<void>;
  emitEntry(
    outputFileName: string,
    printSourceDataCollection: PrintSourceData[]
  ): Promise<void>;
}

export enum DtsModuleType {
  Legacy = "legacy",
  Module = "module",
}

export type PrintSourceData = PrintSourceEnumData | PrintSourceInterfaceData;

export interface PrintSourceDataBase {
  /** 输出目录，输出目录通过 namespace + outDir 决定 */
  outputDir: string;
  moduleType: DtsModuleType;
  fullyQualifiedName: string;
  /** TS 定义所在文件名，生成的 schema 文件需要类型定义 */
  relativeDefinitionFileName?: string;
  /** interface 所有成员，string 表示是原始类型，PrintTypeReference 表示用 ref */
  kind: ts.SyntaxKind;
}

export interface PrintSourceEnumData extends PrintSourceDataBase {
  kind: ts.SyntaxKind.EnumDeclaration;
  members: Array<{ name: string; value: string | number }>;
}

export interface PrintSourceInterfaceData extends PrintSourceDataBase {
  kind: ts.SyntaxKind.InterfaceDeclaration;
  props: Record<string, PrintMemberType>;
}

export interface BasePrintMemberType {
  v: string | PrintTypeReference | "any";
  optional: boolean;
  nullable: boolean;
}

export interface PrintMemberTypeOfArray extends BasePrintMemberType {
  isArray: true;
  latitude: number;
}

export interface PrintMemberTypeOfObject extends BasePrintMemberType {
  isArray: false;
}

export type PrintMemberType = PrintMemberTypeOfArray | PrintMemberTypeOfObject;

export interface PrintTypeReference {
  ref: string;
}

export enum PrinterEmitSchemaType {
  JTD = "jtd",
  JSON = "json",
}

export interface PrinterConfiguration {
  /** output directory */
  outDir?: string;
  emitSchemaType?: PrinterEmitSchemaType;
  prettierConfig?: import("prettier").Options;
}

export interface PrintHost {
  writeFileAsync(fileName: string, content: string): Promise<void>;
  getOutputAbsoluteFileName(printSourceData: PrintSourceData): string;
  normalizeVariableNameFromQualifiedName(fullyQualifiedName: string): string;
}
