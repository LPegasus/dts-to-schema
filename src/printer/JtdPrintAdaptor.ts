import * as path from "path";
import * as ts from "typescript";
import { normalizePath } from "../utils/normalizePath";
import { StringBuilder } from "./StringBuilder";
import {
  DtsModuleType,
  type PrintAdaptor,
  type PrintHost,
  type PrintSourceData,
  PrintMemberTypeOfArray,
  PrintSourceInterfaceData,
  PrintSourceEnumData,
} from "./type";

export class JtdPrintAdaptor implements PrintAdaptor {
  public async emitSchema(printSourceData: PrintSourceData): Promise<void> {
    const { relativeDefinitionFileName, fullyQualifiedName, moduleType, kind } =
      printSourceData;

    const sb = new StringBuilder();
    const outputFileName =
      this.printHost.getOutputAbsoluteFileName(printSourceData);

    /** 引入定义的类型 */
    if (moduleType === DtsModuleType.Legacy) {
      sb.appendLine(
        `/// <reference path="${normalizePath(relativeDefinitionFileName!)}" />`
      );
    } else if (moduleType === DtsModuleType.Module) {
      sb.appendLine(
        `import { ${fullyQualifiedName} } from "${relativeDefinitionFileName}";`
      );
    }

    sb.appendLine();

    if (kind !== ts.SyntaxKind.EnumDeclaration) {
      /** 引入 ajv */
      sb.appendLine(
        `import { JTDSchemaType, SomeJTDSchemaType } from "ajv/dist/core";\n`
      );
    }

    /** export schema */
    sb.appendLine(this._printSchema(printSourceData));

    /** export reference name */
    sb.appendLine(`export const name = "${fullyQualifiedName}";`);

    /** export type */
    sb.appendLine(`export type _T = ${fullyQualifiedName};`);

    return this.printHost
      .writeFileAsync(outputFileName, sb.toString())
      .then(() => {
        this._allOutputSchemaMap.set(outputFileName, printSourceData);
      });
  }

  public async emitEntry(entryFileName: string): Promise<void> {
    const sb = new StringBuilder();
    const definitionsSb = new StringBuilder();
    const schemaCollectionsSb = new StringBuilder();

    sb.appendLine(
      'import type { JTDSchemaType, SomeJTDSchemaType } from "ajv/dist/jtd";'
    );
    definitionsSb.appendLine("const definitions = {");
    schemaCollectionsSb.appendLine("export const SchemaCollections = {");

    this._allOutputSchemaMap.forEach((printSourceData, outputFileName) => {
      const relativeImportPath = normalizePath(
        path.relative(path.dirname(entryFileName), outputFileName)
      ).replace(/(\.d)?\.ts$/, "");

      const variableName =
        this.printHost.normalizeVariableNameFromQualifiedName(
          printSourceData.fullyQualifiedName
        );
      sb.appendLine(
        `import * as ${variableName} from "${relativeImportPath}";`
      );

      definitionsSb.appendLine(
        `[${variableName}.name]: ${variableName}.schema,`
      );

      schemaCollectionsSb.appendLine(
        `[${variableName}.name]: { ref: ${variableName}.name, definitions } as SomeJTDSchemaType as JTDSchemaType<${variableName}._T>,`
      );
    });

    definitionsSb.appendLine("};");
    schemaCollectionsSb.appendLine("};");

    const code = [
      sb.toString(),
      definitionsSb.toString(),
      schemaCollectionsSb.toString(),
    ].join("\n");

    return this.printHost.writeFileAsync(entryFileName, code);
  }

  constructor(protected printHost: PrintHost) {}

  private _allOutputSchemaMap: Map<string, PrintSourceData> = new Map();

  private _printSchema(printSourceData: PrintSourceData): string {
    const sb = new StringBuilder();

    if (printSourceData.kind === ts.SyntaxKind.InterfaceDeclaration) {
      const code = this._appendInterfaceSchemaCode(printSourceData);
      sb.appendLine(code);
    } else if (printSourceData.kind === ts.SyntaxKind.EnumDeclaration) {
      const code = this._appendEnumSchemaCode(printSourceData);
      sb.appendLine(code);
    }

    return sb.toString();
  }

  private _appendInterfaceSchemaCode(
    printSourceData: PrintSourceInterfaceData
  ) {
    const sb = new StringBuilder();
    const keys = Object.keys(printSourceData.props);
    const properties: any = {};
    let optionalProperties: any = undefined;

    keys.forEach((k) => {
      const { isArray, v, optional, nullable } = printSourceData.props[k];
      const data: any = {};
      const typeDescriptor = typeof v === "string" ? getJtdType(v) : v;
      if (isArray) {
        const arrayProp: PrintMemberTypeOfArray = printSourceData.props[
          k
        ] as any;
        let cur = data;
        for (let i = 0; i < arrayProp.latitude - 1; i++) {
          cur = cur.elements = {};
        }
        cur.elements = { type: typeDescriptor };
        if (v === "any") {
          // DO NOTHING
        } else if (typeof v === "string") {
          cur.elements = { type: typeDescriptor };
        } else {
          cur.elements = { ref: v.ref };
        }
      } else {
        if (v === "any") {
          // DO NOTHING
        } else if (typeof v === "string") {
          data.type = typeDescriptor;
        } else {
          data.ref = v.ref;
        }
      }
      if (nullable) {
        data.nullable = true;
      }
      if (optional) {
        if (optionalProperties === null) {
          optionalProperties = {};
        }
        optionalProperties[k] = data;
      } else {
        properties[k] = data;
      }
    });
    sb.append(`export const schema = `);
    sb.appendLine(
      JSON.stringify({ properties, optionalProperties }, undefined, 2) +
        ` as SomeJTDSchemaType as JTDSchemaType<${printSourceData.fullyQualifiedName}>;`
    );
    return sb.toString();
  }

  /**
   * 打印 enum 取值含义
   *
   * @param printSourceData - PrintSourceEnumData
   */
  private _appendEnumSchemaCode(printSourceData: PrintSourceEnumData): string {
    const sb = new StringBuilder();
    if (printSourceData.members.length > 0) {
      sb.appendLine(`/**`);
      for (const member of printSourceData.members) {
        sb.appendLine(` * - ${member.value}: ${member.name}`);
      }
      sb.appendLine(` */`);
    }
    sb.appendLine('export const schema = { type: "int32" };');
    return sb.toString();
  }
}

function getJtdType(typeIdentifier: string) {
  switch (typeIdentifier) {
    case "string":
      return "string";
    case "number":
      return "float64";
    case "boolean":
      return "boolean";
    case "any":
      return {};
  }

  throw new Error('Unknown primitive type "' + typeIdentifier + '"');
}
