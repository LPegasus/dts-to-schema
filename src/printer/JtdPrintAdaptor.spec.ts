import { JtdPrintAdaptor } from "./JtdPrintAdaptor";
import { pascalCase } from "change-case";
import { DtsModuleType } from "./type";
import * as ts from "typescript";

function getPrinter(writeFileAsync: any) {
  const printer = new JtdPrintAdaptor({
    writeFileAsync,
    getOutputAbsoluteFileName: (printSourceData) => {
      return printSourceData.fullyQualifiedName + ".schema.ts";
    },
    normalizeVariableNameFromQualifiedName: (qualifiedName) => {
      return pascalCase(qualifiedName);
    },
  });
  return printer;
}

const _arrayBase = {
  fullyQualifiedName: "ArrayBase",
  kind: ts.SyntaxKind.InterfaceDeclaration,
  moduleType: DtsModuleType.Legacy,
  outputDir: "",
  relativeDefinitionFileName: "../dts/ArrayBase.d.ts",
  props: {
    arr1: {
      isArray: true,
      nullable: false,
      optional: false,
      latitude: 1,
      v: "string",
    },
    arr2: {
      isArray: true,
      nullable: false,
      optional: true,
      v: "number",
      latitude: 3,
    },
    arr3: {
      isArray: true,
      nullable: true,
      optional: false,
      v: { ref: "Some.Other.Type" },
      latitude: 2,
    },
  },
} as const;

const _base = {
  fullyQualifiedName: "Base",
  kind: ts.SyntaxKind.InterfaceDeclaration,
  moduleType: DtsModuleType.Legacy,
  outputDir: "",
  relativeDefinitionFileName: "../dts/Base.d.ts",
  props: {
    foo: {
      isArray: false,
      nullable: false,
      optional: false,
      v: "string",
    },
    optionalField: {
      isArray: false,
      nullable: false,
      optional: true,
      v: "number",
    },
  },
} as const;

describe("JtdPrintAdaptor", () => {
  test("emitSchema base", async () => {
    const writeFileAsync = jest.fn((arg0: string, arg1: string) =>
      Promise.resolve()
    );
    const printer = getPrinter(writeFileAsync);

    await printer.emitSchema(_base);

    // @ts-ignore
    expect(writeFileAsync.mock.calls[0][1]).toMatchSnapshot();
  });

  test("emitSchema with array", async () => {
    const writeFileAsync = jest.fn((arg0: string, arg1: string) =>
      Promise.resolve()
    );
    const printer = getPrinter(writeFileAsync);

    await printer.emitSchema({
      fullyQualifiedName: "ArrayBase",
      kind: ts.SyntaxKind.InterfaceDeclaration,
      moduleType: DtsModuleType.Legacy,
      outputDir: "",
      relativeDefinitionFileName: "../dts/ArrayBase.d.ts",
      props: {
        arr1: {
          isArray: true,
          nullable: false,
          optional: false,
          latitude: 1,
          v: "string",
        },
        arr2: {
          isArray: true,
          nullable: false,
          optional: true,
          v: "number",
          latitude: 3,
        },
        arr3: {
          isArray: true,
          nullable: true,
          optional: false,
          v: { ref: "Some.Other.Type" },
          latitude: 2,
        },
      },
    });

    expect(writeFileAsync.mock.calls[0][1]).toMatchSnapshot();
  });

  test("emitEntry", async () => {
    const writeFileAsync = jest.fn((arg0: string, arg1: string) =>
      Promise.resolve()
    );
    const printer = getPrinter(writeFileAsync);
    const p2 = printer.emitSchema(_arrayBase);
    const p1 = printer.emitSchema(_base);
    await Promise.all([p1, p2]);

    await printer.emitEntry("/entry.ts");

    expect(writeFileAsync).toHaveBeenCalledTimes(3);

    expect(writeFileAsync.mock.calls[2][1]).toMatchSnapshot();
  });
});
