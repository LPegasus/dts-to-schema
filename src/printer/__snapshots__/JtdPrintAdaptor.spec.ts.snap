// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JtdPrintAdaptor emitEntry 1`] = `
"import type { JTDSchemaType, SomeJTDSchemaType } from \\"ajv/dist/jtd\\";
import * as ArrayBase from \\"../ts-json-schema/ArrayBase.schema.ts\\";
import * as Base from \\"../ts-json-schema/Base.schema.ts\\";
const definitions = {
[ArrayBase.name]: ArrayBase.schema,
[Base.name]: Base.schema,
};
export const SchemaCollections = {
[ArrayBase.name]: { ref: ArrayBase.name, definitions } as SomeJTDSchemaType as JTDSchemaType<ArrayBase._T>,
[Base.name]: { ref: Base.name, definitions } as SomeJTDSchemaType as JTDSchemaType<Base._T>,
};
"
`;

exports[`JtdPrintAdaptor emitSchema base 1`] = `
"/// <reference path=\\"../dts/Base.d.ts\\" />
import { JTDSchemaType, SomeJTDSchemaType } from \\"ajv/dist/core\\";
export const schema = {
  \\"properties\\": {
    \\"foo\\": {
      \\"type\\": \\"string\\"
    }
  },
  \\"optionalProperties\\": {
    \\"optionalField\\": {
      \\"type\\": \\"float64\\"
    }
  }
} as SomeJTDSchemaType as JTDSchemaType<Base>;

export const name = \\"Base\\";
export type _T = Base;
"
`;

exports[`JtdPrintAdaptor emitSchema with array 1`] = `
"/// <reference path=\\"../dts/ArrayBase.d.ts\\" />
import { JTDSchemaType, SomeJTDSchemaType } from \\"ajv/dist/core\\";
export const schema = {
  \\"properties\\": {
    \\"arr1\\": {
      \\"elements\\": \\"string\\"
    },
    \\"arr3\\": {
      \\"elements\\": {
        \\"elements\\": {
          \\"ref\\": \\"Some.Other.Type\\"
        }
      },
      \\"nullable\\": true
    }
  },
  \\"optionalProperties\\": {
    \\"arr2\\": {
      \\"elements\\": {
        \\"elements\\": {
          \\"elements\\": \\"float64\\"
        }
      }
    }
  }
} as SomeJTDSchemaType as JTDSchemaType<ArrayBase>;

export const name = \\"ArrayBase\\";
export type _T = ArrayBase;
"
`;
